src/group_theory/order_of_element.lean

```diff
@@ -345,41 +506,84 @@ begin
 finset.le_card_of_inj_on_range ((^) a)
   (assume n _, finset.mem_univ _)
   (assume i hi j hj, pow_injective_of_lt_order_of a hi hj)
 
 /-- This is the same as `order_of_pos' but with one fewer explicit assumption since this is
   automatic in case of a finite cancellative monoid.-/
-lemma order_of_pos (a : α) : 0 < order_of a :=
+lemma order_of_pos (a : α) : 0 < order_of a := order_of_pos' (exists_pow_eq_one a)
+
+/-- This is the same as `add_order_of_pos' but with one fewer explicit assumption since this is
+  automatic in case of a finite cancellative additive monoid.-/
+lemma add_order_of_pos (x : H) : 0 < add_order_of x :=
 begin
-  apply order_of_pos',
-  cases exists_pow_eq_one a with w hw,
-  cases hw with hw1 hw2,
-  exact ⟨w, hw1, hw2⟩,
+  rw ← order_of_of_add_eq_add_order_of,
+  exact order_of_pos _,
 end
 
+attribute [to_additive add_order_of_pos] order_of_pos
+
 variables {n : ℕ}
 
 open nat
 
 /-- This is the same as `order_of_pow'` and `order_of_pow''` but with one assumption less which is
-automatic in the case of a finite cancellative group.-/
+automatic in the case of a finite cancellative monoid.-/
 lemma order_of_pow (a : α) :
-  order_of (a ^ n) = order_of a / gcd (order_of a) n :=
+  order_of (a ^ n) = order_of a / gcd (order_of a) n := order_of_pow'' _ _ (exists_pow_eq_one _)
+
+/-- This is the same as `add_order_of_nsmul'` and `add_order_of_nsmul` but with one assumption less
+which is automatic in the case of a finite cancellative additive monoid. -/
+lemma add_order_of_nsmul (x : H) :
+  add_order_of (n •ℕ x) = add_order_of x / gcd (add_order_of x) n :=
 begin
-  apply order_of_pow'',
-  cases exists_pow_eq_one a with w hw,
-  cases hw with hw1 hw2,
-  exact ⟨w, hw1, hw2⟩
+  rw [← order_of_of_add_eq_add_order_of, of_add_nsmul],
+  exact order_of_pow _,
 end
 
+attribute [to_additive add_order_of_nsmul] order_of_pow
+
+lemma mem_multiples_iff_mem_range_add_order_of [decidable_eq H] {x x' : H} :
+  x' ∈ add_submonoid.multiples x ↔
+  x' ∈ (finset.range (add_order_of x)).image ((•ℕ x) : ℕ → H)  :=
+finset.mem_range_iff_mem_finset_range_of_mod_eq' (add_order_of_pos x)
+  (assume i, nsmul_eq_mod_add_order_of.symm)
+
+@[to_additive mem_multiples_iff_mem_range_add_order_of]
 lemma mem_powers_iff_mem_range_order_of [decidable_eq α] {a a' : α} :
   a' ∈ submonoid.powers a ↔ a' ∈ (finset.range (order_of a)).image ((^) a : ℕ → α) :=
 finset.mem_range_iff_mem_finset_range_of_mod_eq' (order_of_pos a)
   (assume i, pow_eq_mod_order_of.symm)
 
+noncomputable instance decidable_multiples [decidable_eq H] {x : H} :
+  decidable_pred (add_submonoid.multiples x : set H) :=
+begin
+  assume x',
+  apply decidable_of_iff' (x' ∈ (finset.range (add_order_of x)).image (•ℕ x)),
+  exact mem_multiples_iff_mem_range_add_order_of,
+end
+
+@[to_additive decidable_multiples]
 noncomputable instance decidable_powers [decidable_eq α] :
   decidable_pred (submonoid.powers a : set α) :=
```

Comment:
`noncomputable instance`s of `decidable` is a very weird idea to me: why not just use `classical`? On the other hand, it was already there for `powers`, so presumably there is a good reason to have it?

Reply:
Removing the instance makes `order_eq_card_powers`, `order_eq_card_gpowers` and their additive versions not typecheck. I'm happy to remove these instances if this can be achieved in some other way. Still not quite familiar with decidability.

Reply:
That seems like a good enough reason to keep them around in this PR. I'll see what I can do with the instances myself later today.